### PR TITLE
Make ObjectStoreProvider fallible (return `Result` rather than `Option`)

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

re https://github.com/apache/arrow-datafusion/issues/3424

 # Rationale for this change

https://github.com/apache/arrow-datafusion/pull/3540#issuecomment-1254598498

# What changes are included in this PR?

Makes `ObjectStoreProvider::get_by_url` fallible and clarifies some docs

# Are there any user-facing changes?

### CODE REVIEW

**Review by alamb — APPROVED** [review]
cc @yahoNanJing and @thinkharderdev  for the next ballista upgrade

**Comment by alamb** [comment]
Thanks @tustvold

File: datafusion/core/src/datasource/object_store.rs [hunk 1]
```diff
@@ -83,9 +83,8 @@ impl std::fmt::Display for ObjectStoreUrl {
 
 /// Object store provider can detector an object store based on the url
 pub trait ObjectStoreProvider: Send + Sync + 'static {
-    /// Detector a suitable object store based on its url if possible
-    /// Return the key and object store
-    fn get_by_url(&self, url: &Url) -> Option<Arc<dyn ObjectStore>>;
+    /// Return an ObjectStore for the provided url based on its scheme and authority
+    fn get_by_url(&self, url: &Url) -> Result<Arc<dyn ObjectStore>>;
```
Comment (yahoNanJing) [inline datafusion/core/src/datasource/object_store.rs:87]: Better to change it to be Result<Option<Arc<dyn ObjectStore>>>

**Comment by yahoNanJing** [comment]
Thanks @alamb for this info. I think better to change the return type to be 
```
Result<Option<Arc<dyn ObjectStore>>>
```

**Comment by alamb** [comment]
Follow on PR and discussion for `Result<Option>` is https://github.com/apache/arrow-datafusion/pull/3595

**Pull request merged**